### ISSUE TITLE
Extract the testing framework used in nim-chronicles into a more reusable library

### ISSUE THREAD

**Issue by zah** [status-im/nim-chronicles#71]
Chronicles includes a testing framework similar in spirit to Nim's testament. Instead of simply running unit tests, it allow you to compile Nim programs with various compilation flags, then to run the resulting binary with various command-line params and to compare the output of the program with a known good result.

Each test case is described in a NimCfg-parsable text file with the "test" extension. Here is an example:
https://github.com/status-im/nim-chronicles/blob/master/tests/loglevel/no_level.test

At the top of the file, you see a number of options specified. Some of them are predefined options expected by the test runner, while all other are considered nim defines (i.e. `-d:opt=value`). 

The [testrunner](https://github.com/status-im/nim-chronicles/blob/master/tests/testrunner.nim) program will scan a directory for such files and it will execute them in turn.

Goals of this bounty:

1) Extract the test runner code into a reusable repo (e.g. a new nim-testutils package or in nim-stew/stew/testing). Our goal is to use this package in other libraries such as [nim-confutils](https://github.com/status-im/nim-confutils).

2) Add the ability to specify command-line options in each test execution. Ideally, the test file corresponds to a single build of the target binary, but it allows you to specify multiple test executions.

3) Add a command that allows you to re-generate the output section in a test file by running the specified program and trusting its output. This is useful when you intentionally modify the behavior of the tested program and now it produces a different output. It's expected that the programmer will review the result in a diff viewer after this operation.

4) Stretch goal: Allow the test runner to execute the tests in parallel.

**Comment by disruptek** [comment]
This is basically just `testrunner` at the moment.
https://github.com/disruptek/testutils

I'll continue down the list tomorrow.

**Comment by disruptek** [comment]
### concurrency

My plan is for each combination of compilation arguments to produce a unique compilation with its own nimcache, etc.  Each such compilation gets a thread, and we'll build everything first; compilation is a prerequisite of execution, after all.  I do this in openapi and it parallelizes just fine.

After compilation is complete, we will run as many tests as we can for each program, in series.  This way, you have a pretty good idea of the resources available or consumed at any point in the test process, and you can use that to define a threadcount for the executions that makes sense.

You can affect order by naming your files with `00` and `10` prefixes.  Maybe it would be clever to, by default, run the most-recently-modified tests first.

## choosing to update output
When we review the result of an updated test, do we use `diff` or `git-configured difftool` or similar?  I noticed that there are no Windows tests.  Are Windows users out of luck?   Or do we implement a simple on-board diff renderer?

**Comment by zah** [comment]
> Re mini-diff

What I meant is that the developer could easily use his existing tooling (e.g `git diff`) to review the test files modifications after the testrunner has updated them. Where a mini-diff engine would be useful is highlighting what is different in the test result from the expected output. It would be unwise to spent too much time on this though, I see it as a worthwhile addition only if we reuse an existing diffing engine. I know that testament has one, so perhaps you can look into how much work will be to extract it from there.

**Comment by zah** [comment]
For the [completion of the goals](https://github.com/status-im/nim-testutils/commits?author=disruptek) stated in this bounty, @disruptek has been awarded 560 USD (in DAI):
https://etherscan.io/tx/0x578968a6eb18faf6be93a4dbfe9c79067122f79a7b7d95bdd8e2aee26cac5555
